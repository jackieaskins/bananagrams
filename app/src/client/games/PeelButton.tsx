import {
  Button,
  ButtonProps,
  IconButton,
  Text,
  Tooltip,
} from "@chakra-ui/react";
import { useMemo } from "react";
import { FaPlay } from "react-icons/fa6";
import { useGame } from "./GameContext";
import { isValidConnectedBoard } from "@/client/boards/validate";
import { useCurrentPlayer } from "@/client/redesign/useCurrentPlayer";
import { PlayerStatus } from "@/types/player";

export interface PeelButtonProps extends ButtonProps {
  hideText?: boolean;
}

export default function PeelButton({
  hideText,
  ...buttonProps
}: PeelButtonProps): JSX.Element {
  const {
    gameInfo: { bunch, players },
    handlePeel,
  } = useGame();
  const { board, hand } = useCurrentPlayer();

  const activePlayers = useMemo(
    () => players.filter(({ status }) => status === PlayerStatus.READY),
    [players],
  );

  const canPeel = hand.length === 0 && isValidConnectedBoard(board);
  const peelWinsGame = bunch.length < activePlayers.length;

  const peelButtonHint = useMemo(() => {
    if (!canPeel) {
      return "You must have a valid connected board to peel";
    }

    if (peelWinsGame) {
      return "Win the game!";
    }

    return "Get a new tile and send one to everyone else";
  }, [canPeel, peelWinsGame]);

  const commonProps = useMemo(
    () => ({
      ...buttonProps,
      colorScheme: "blue",
      onClick: handlePeel,
      isDisabled: !canPeel,
    }),
    [buttonProps, canPeel, handlePeel],
  );

  return (
    <Tooltip label={<Text textAlign="center">{peelButtonHint}</Text>} hasArrow>
      {hideText ? (
        <IconButton
          {...commonProps}
          icon={<FaPlay />}
          aria-label={peelWinsGame ? "Bananas!" : "Peel!"}
        />
      ) : (
        <Button {...commonProps} leftIcon={<FaPlay />}>
          {peelWinsGame ? "Bananas!" : "Peel!"}
        </Button>
      )}
    </Tooltip>
  );
}
