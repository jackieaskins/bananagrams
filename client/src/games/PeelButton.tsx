import { Tooltip } from "@mui/material";
import Button from "../buttons/Button";

type PeelButtonProps = {
  canPeel: boolean;
  handlePeel: () => void;
  peelWinsGame: boolean;
};

const PeelButton: React.FC<PeelButtonProps> = ({
  canPeel,
  handlePeel,
  peelWinsGame,
}) => {
  const getPeelButtonHint = (): string => {
    if (!canPeel) {
      return "You must have a valid connected board to peel";
    }

    if (peelWinsGame) {
      return "Win the game!";
    }

    return "Get a new tile and send one to everyone else";
  };

  return (
    <Tooltip title={getPeelButtonHint()}>
      <span>
        <Button size="large" fullWidth onClick={handlePeel} disabled={!canPeel}>
          {peelWinsGame ? "Bananas!" : "Peel!"}
        </Button>
      </span>
    </Tooltip>
  );
};

export default PeelButton;
