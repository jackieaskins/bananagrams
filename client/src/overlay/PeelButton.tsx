import { Button } from 'antd';
import { memo, useMemo } from 'react';

import { isValidConnectedBoard } from '../boards/validate';
import {
  useCurrentBoard,
  useCurrentHand,
  useGameBunchCount,
  useGamePlayers,
} from '../game/stateHooks';
import { peel } from '../socket';

const PeelButton = (): JSX.Element => {
  const hand = useCurrentHand();
  const board = useCurrentBoard();
  const players = useGamePlayers();
  const bunchCount = useGameBunchCount();

  const canPeel = useMemo(
    () => hand.length === 0 && isValidConnectedBoard(board),
    [hand, board]
  );
  const peelWinsGame = useMemo(() => bunchCount < players.length, [
    bunchCount,
    players,
  ]);

  return useMemo(
    () => (
      <Button type="primary" disabled={!canPeel} onClick={peel} block>
        {peelWinsGame ? 'Bananas' : 'Peel'}
      </Button>
    ),
    [canPeel, peelWinsGame]
  );
};

export default memo(PeelButton);
