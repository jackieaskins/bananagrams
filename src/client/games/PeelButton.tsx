import {
  Button,
  ButtonProps,
  IconButton,
  Text,
  Tooltip,
} from "@chakra-ui/react";
import { useMemo } from "react";
import { FaPlay } from "react-icons/fa6";
import { PlayerStatus } from "../../types/player";
import { isValidConnectedBoard } from "../boards/validate";
import { useCurrentPlayer } from "../redesign/useCurrentPlayer";
import { useGame } from "./GameContext";

export interface PeelButtonProps extends ButtonProps {
  hideText?: boolean;
}

export default function PeelButton({
  hideText,
  ...buttonProps
}: PeelButtonProps): JSX.Element {
  const {
    gameInfo: { bunch, players },
    handlePeel,
  } = useGame();
  const { board, hand } = useCurrentPlayer();

  const activePlayers = useMemo(
    () => players.filter(({ status }) => status === PlayerStatus.READY),
    [players],
  );

  const canPeel = hand.length === 0 && isValidConnectedBoard(board);
  const peelWinsGame = bunch.length < activePlayers.length;

  const peelButtonHint = useMemo(() => {
    if (!canPeel) {
      return "You must have a valid connected board to peel";
    }

    if (peelWinsGame) {
      return "Win the game!";
    }

    return "Get a new tile and send one to everyone else";
  }, [canPeel, peelWinsGame]);

  const commonProps = useMemo(
    () => ({
      ...buttonProps,
      colorScheme: "blue",
      onClick: handlePeel,
      isDisabled: !canPeel,
    }),
    [buttonProps, canPeel, handlePeel],
  );

  return (
    <Tooltip label={<Text textAlign="center">{peelButtonHint}</Text>} hasArrow>
      {hideText ? (
        <IconButton
          {...commonProps}
          icon={<FaPlay />}
          aria-label={peelWinsGame ? "Bananas!" : "Peel!"}
        />
      ) : (
        <Button {...commonProps} leftIcon={<FaPlay />}>
          {peelWinsGame ? "Bananas!" : "Peel!"}
        </Button>
      )}
    </Tooltip>
  );
}
