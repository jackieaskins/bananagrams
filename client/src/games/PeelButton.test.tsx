import { Button } from "@chakra-ui/react";
import { shallow } from "enzyme";
import PeelButton from "./PeelButton";

describe("<PeelButton />", () => {
  const renderComponent = (propOverrides = {}) =>
    shallow(
      <PeelButton
        canPeel
        handlePeel={jest.fn().mockName("handlePeel")}
        peelWinsGame={false}
        {...propOverrides}
      />,
    );

  test("renders properly", () => {
    expect(renderComponent()).toMatchSnapshot();
  });

  describe("tooltip text", () => {
    test("gives help on how to disable button if cannot peel", () => {
      expect(
        renderComponent({ canPeel: false }).props().label.props.children,
      ).toMatch("You must have a valid connected board to peel");
    });

    test("shows win game text when can peel and peel wins game", () => {
      expect(
        renderComponent({ canPeel: true, peelWinsGame: true }).props().label
          .props.children,
      ).toBe("Win the game!");
    });

    test("shows peel description when can peel but peel does not win game", () => {
      expect(
        renderComponent({ canPeel: true, peelWinsGame: false }).props().label
          .props.children,
      ).toBe("Get a new tile and send one to everyone else");
    });
  });

  describe("button text", () => {
    test("shows bananas when peel wins game", () => {
      expect(
        renderComponent({ peelWinsGame: true }).find(Button).props().children,
      ).toBe("Bananas!");
    });

    test("shows peel when peel does not win game", () => {
      expect(
        renderComponent({ peelWinsGame: false }).find(Button).props().children,
      ).toBe("Peel!");
    });
  });

  describe("disabled state", () => {
    test("is disabled when cannot peel", () => {
      expect(
        renderComponent({ canPeel: false }).find(Button).props().isDisabled,
      ).toBe(true);
    });

    test("is not disabled when can peel", () => {
      expect(
        renderComponent({ canPeel: true }).find(Button).props().isDisabled,
      ).toBe(false);
    });
  });
});
