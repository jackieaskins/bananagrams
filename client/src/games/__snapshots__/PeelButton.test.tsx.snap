// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PeelButton /> renders properly 1`] = `
<ForwardRef(Tooltip)
  title="Get a new tile and send one to everyone else"
>
  <span>
    <Button
      disabled={false}
      fullWidth={true}
      onClick={[MockFunction handlePeel]}
      size="large"
    >
      Peel!
    </Button>
  </span>
</ForwardRef(Tooltip)>
`;
